### PR TITLE
feat(manager/terraform): support range strategy update lockfile

### PR DESCRIPTION
**Pull request by secustor**
## Changes:
Adapt the Terraform manager and Hashicorp versioning to allow the rangeStrategy `update-lockfile` 

## Context:
https://github.com/renovatebot/renovate/discussions/11701

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

https://github.com/secustor/terraform-pessimistic-version-constraint

### CODE REVIEW

File: lib/manager/terraform/extract.spec.ts [hunk 1]
```diff
@@ -1,32 +1,65 @@
...
+  cacheDir: join('/tmp/cache'),
+};
+
+// auto-mock fs
+jest.mock('../../util/fs');
 
 describe('manager/terraform/extract', () => {
+  beforeEach(() => {
+    jest.resetAllMocks();
+    jest.resetModules();
```
Comment (viceice) [inline lib/manager/terraform/extract.spec.ts:28]: I think we don't need this.
```suggestion
```

File: lib/manager/terraform/util.ts [hunk 2]
```diff
@@ -42,3 +45,32 @@ export function checkIfStringIsPath(path: string): boolean {
   const match = pathStringRegex.exec(path);
   return !!match;
 }
+
+export function massageProviderLookupName(dep: PackageDependency): void {
```
Comment (viceice) [inline lib/manager/terraform/util.ts:49]: Do we really need this?

File: lib/manager/terraform/util.ts [hunk 3]
```diff
@@ -42,3 +45,32 @@ export function checkIfStringIsPath(path: string): boolean {
...
+  }
+  /* eslint-enable no-param-reassign */
+}
+
+export function getLockedVersion(
+  dep: PackageDependency,
+  locks: ProviderLock[]
+): string {
+  const foundLock = locks.find((lock) => {
+    const depRegistryUrl = dep.registryUrls
```
Comment (viceice) [inline lib/manager/terraform/util.ts:65]: Move this outside loop. 😉

**Review by viceice — COMMENTED** [review]
Needs code coverage

File: lib/manager/terraform/util.ts (hunk 2 above)
Comment (secustor) [reply lib/manager/terraform/util.ts:49]: I use it to encapsulate the massaging code and I reuse it now in the `updateArtifacts` function

File: lib/manager/terraform/extract.spec.ts [hunk 4]
```diff
@@ -1,32 +1,63 @@
...
-      const res = extractPackageFile(helm);
+    it('extract helm releases', async () => {
+      const res = await extractPackageFile(helm, 'helm.tf', {});
       expect(res).toMatchSnapshot();
       expect(res.deps).toHaveLength(6);
       expect(res.deps.filter((dep) => dep.skipReason)).toHaveLength(2);
     });
+
+    it('update lockfile constraints with range strategy update-lockfile', async () => {
+      fs.readLocalFile.mockResolvedValueOnce(lockedVersionLockfile as any);
```
Comment (viceice) [inline lib/manager/terraform/extract.spec.ts:50]:
```suggestion
      fs.readLocalFile.mockResolvedValueOnce(lockedVersionLockfile);
```

I think those `any` casts are no longer required

File: lib/manager/terraform/extract.ts [hunk 5]
```diff
@@ -1,6 +1,8 @@
 import { logger } from '../../logger';
+import { ExtractConfig } from '../types';
```
Comment (viceice) [inline lib/manager/terraform/extract.ts:2]:
```suggestion
import type { ExtractConfig } from '../types';
```

File: lib/manager/terraform/lockfile/index.ts [hunk 6]
```diff
@@ -85,30 +86,23 @@ export async function updateArtifacts({
         ['provider', 'required_provider'].includes(dep.depType)
       );
       for (const dep of providerDeps) {
-        const lookupName = dep.lookupName ?? dep.depName;
+        massageProviderLookupName(dep);
```
Comment (viceice) [inline lib/manager/terraform/lockfile/index.ts:89]: isn't this already done in extraction phase? Do we need to make this twice? 🤔

Comment (secustor) [reply lib/manager/terraform/lockfile/index.ts:89]: Yes, should be. I have added this to be for sure backwards compatible.

Comment (viceice) [reply lib/manager/terraform/lockfile/index.ts:89]: OK, so we should cleanup in a new PR. Looks pretty ugly now.

File: lib/manager/terraform/extract.spec.ts [hunk 7]
```diff
@@ -1,32 +1,63 @@
...
+  });
   describe('extractPackageFile()', () => {
-    it('returns null for empty', () => {
-      expect(extractPackageFile('nothing here')).toBeNull();
+    it('returns null for empty', async () => {
+      expect(await extractPackageFile('nothing here', '1.tf', {})).toBeNull();
     });
-    it('extracts', () => {
-      const res = extractPackageFile(tf1);
+    it('extracts', async () => {
```
Comment (viceice) [inline lib/manager/terraform/extract.spec.ts:33]:
```suggestion

    it('extracts', async () => {
```

File: lib/manager/terraform/extract.spec.ts [hunk 8]
```diff
@@ -1,32 +1,63 @@
...
+      const res = await extractPackageFile(tf1, '1.tf', {});
       expect(res).toMatchSnapshot();
       expect(res.deps).toHaveLength(46);
       expect(res.deps.filter((dep) => dep.skipReason)).toHaveLength(8);
     });
-    it('returns null if only local deps', () => {
-      expect(extractPackageFile(tf2)).toBeNull();
+    it('returns null if only local deps', async () => {
```
Comment (viceice) [inline lib/manager/terraform/extract.spec.ts:39]:
```suggestion

    it('returns null if only local deps', async () => {
```

File: lib/manager/terraform/extract.spec.ts [hunk 9]
```diff
@@ -1,32 +1,63 @@
...
+      expect(await extractPackageFile(tf2, '2.tf', {})).toBeNull();
     });
-    it('extract helm releases', () => {
-      const res = extractPackageFile(helm);
+    it('extract helm releases', async () => {
```
Comment (viceice) [inline lib/manager/terraform/extract.spec.ts:42]:
```suggestion

    it('extract helm releases', async () => {
```

**Review by viceice — COMMENTED** [review]
some small cleanup

**Review by viceice — APPROVED** [review]

**Pull request merged**